src/shared/switch-root.c

```diff
@@ -26,41 +26,44 @@
 #include "user-util.h"
 
 int switch_root(const char *new_root,
-                const char *old_root_after, /* path below the new root, where to place the old root after the transition */
-                bool unmount_old_root,
-                unsigned long mount_flags) {  /* MS_MOVE or MS_BIND */
+                const char *old_root_after,   /* path below the new root, where to place the old root after the transition; may be NULL to unmount it */
+                unsigned long mount_flags) {  /* MS_MOVE or MS_BIND used for /proc/, /dev/, /run/, /sys/ */
 
+        _cleanup_close_ int old_root_fd = -EBADF, new_root_fd = -EBADF;
         _cleanup_free_ char *resolved_old_root_after = NULL;
-        _cleanup_close_ int old_root_fd = -EBADF;
-        int r;
+        int r, istmp;
 
         assert(new_root);
-        assert(old_root_after);
 
         if (path_equal(new_root, "/"))
                 return 0;
 
         /* Check if we shall remove the contents of the old root */
-        old_root_fd = open("/", O_RDONLY | O_CLOEXEC | O_DIRECTORY);
+        old_root_fd = open("/", O_DIRECTORY|O_CLOEXEC);
         if (old_root_fd < 0)
                 return log_error_errno(errno, "Failed to open root directory: %m");
-        r = fd_is_temporary_fs(old_root_fd);
-        if (r < 0)
-                return log_error_errno(r, "Failed to stat root directory: %m");
-        if (r > 0)
+
+        istmp = fd_is_temporary_fs(old_root_fd);
+        if (istmp < 0)
+                return log_error_errno(istmp, "Failed to stat root directory: %m");
+        if (istmp > 0)
                 log_debug("Root directory is on tmpfs, will do cleanup later.");
```

Comment:
What happens if the current root is on a tmpfs? IIRC it's one of the schemes we support, with /usr and /var and /etc as separate partitions?

Reply:
ie, maybe it should also check that the old root is an initrd explicitly?

Reply:
this PR does not make any change on this.

But frankly I think we should just say that "switch-root" is for initrd → host transitions, and maybe block the whole operation if /etc/initrd-release doesn't exist.

And then advertise the new "renew" boot as the way for everything else.

But given this PR doesn't actually touch this logic at all, let's do this separately.

Reply:
ok, but add it to the TODO or something so we don't forget - as this will delete everything, and if that happens to root because it happens to be a tmpfs, it's not going to be nice

Reply:
will add to todo list as part of #27435, since i think tightening the rules can only be done once userspace reboot is implemented, to give people a clear escape hatch for transitioning into new root fs